File: app/src/main/java/com/thedistantblue/triaryapp/mainscreen/power/detail/exercise/PowerExerciseViewModel.java
package com.thedistantblue.triaryapp.mainscreen.power.detail.exercise;

import androidx.databinding.ObservableField;
import androidx.lifecycle.ViewModel;

import com.thedistantblue.triaryapp.entities.base.Exercise;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class PowerExerciseViewModel extends ViewModel {

    private final PowerExerciseFragment powerExerciseFragment;
    public final ObservableField<String> powerExerciseName = new ObservableField<>();

    public void createExercise() {
        Exercise.Builder exerciseBuilder = new Exercise.Builder().setName(powerExerciseName.get());
        powerExerciseFragment.createExercise(exerciseBuilder);
    }

}

File: app/src/main/kotlin/com/thedistantblue/triaryapp/mainscreen/power/detail/exercise/compose/PowerExerciseViewModel.kt
package com.thedistantblue.triaryapp.mainscreen.power.detail.exercise.compose

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import com.thedistantblue.triaryapp.database.room.dao.ExerciseDao
import com.thedistantblue.triaryapp.entities.base.Exercise

class PowerExerciseViewModel(
        private val exerciseDao: ExerciseDao
): ViewModel() {
    val uiState: MutableState<Exercise> = mutableStateOf(Exercise())
    private var currentExercise = uiState.value

    fun updateExerciseName(name: String) {
        currentExercise = Exercise(currentExercise.exerciseId,
                                   currentExercise.trainingId,
                                   name,
                                   currentExercise.description)
        uiState.value = currentExercise
    }

    fun updateExerciseDescription(description: String) {
        currentExercise = Exercise(currentExercise.exerciseId,
                                   currentExercise.trainingId,
                                   currentExercise.name,
                                   description)
        uiState.value = currentExercise
    }

    fun getExercise(exerciseId: String) {
        currentExercise.exerciseId?.let {
            if (it.toString() != exerciseId) {
                exerciseDao.findById(exerciseId).subscribe { exercise ->
                    currentExercise = exercise
                    uiState.value = currentExercise
                }
            }
        } ?: run {
            exerciseDao.findById(exerciseId).subscribe { exercise ->
                currentExercise = exercise
                uiState.value = currentExercise
            }
        }
    }

    fun createExercise(exercise: Exercise) {
        exerciseDao.create(exercise).subscribe()
    }

    fun saveExercise(exercise: Exercise) {
        exerciseDao.save(exercise).subscribe()
    }

}